### DISCUSSION TITLE
Info concettuali: integrazioni componenti termici (PdC, termoconvettori) ed entità modbus

### DISCUSSION THREAD

**Discussion by crug80** [indomus/forum#2767] [GENERALE - Discussioni generiche]
Salve, 
sto integrando i componenti relativi all'impianto termico in HA tramite l'integrazione modbus. Al momento mi sto limitando ai ventilconvettori, poi passerò alla pompa di calore. 
L'integrazione "funziona", ma vorrei raffinarla e ci sono alcuni concetti che mi sfuggono. 
Concettualmente per ogni termoconvettore ho creato le entità sensori per leggere le temperature (ambiente + piastra), un'entità climates per impostare i setpoint, la temperatura target ed accendere e spegnerli ae 4 entità fan, una per ogni velocità disponibile. 
Il tutto, come detto funziona, eccetto per il fatto che i miei termoconvettori hanno registri dedicati alla temperatura target per le diverse modalità, quindi uno per quando siamo in caldo, una per il freddo ed uno per auto.. 
Con l'approccio che ho seguito al momento riesco ad impostare solo un registro temperatura che è quello per il caldo. 
Le miei domande sono: è giusto il mio approccio? Quindi tante entità diverse per comandare un singolo fancoil? Se sì, come posso gestire all'interno del climates i diversi registri temperatura? (template?!)

Riguardo la PdC, invece, il componente climates è decisamente limitato, poiché oltre alle diverse temperature e velocità che posso ricavarmi come sensori, ho diversi setpoint e funzionalità che potrei gestire (sorgente suplementare del sistema ibrido, limitazione della potenza assorbita etc.)
Da sviluppatore software (altri ambiti, altri settori ed altri linguaggi) direi che dovrei crearmi un mio componente, ma sinceramente ho difficoltà nel reperire le giuste informazioni da cui partire e, soprattutto, a capire se questo sforzo è obbligato o posso cavarmela diversamente e più velocemente. 
Qualcuno che ha voglia di ragionare con me?

**Comment by Marcopal71** [comment]
Buongiorno, una domanda se possibile, riguardo la PdC hai già l'elenco dei registri da utilizzare? io ho una Italtherm ma non riesco a trovare nulla, sul pannello operatore sono presenti i morsetti A - B per una connessione modbus ma anche interpellando la casa madre non danno nulla.

Grazie
Marco

**Reply by crug80** [reply]
Io ho già tutto, poiché all'epoca dell'acquisto mi ero già orientato su modelli che avevano l'opzione di interfacciamento con sistemi bms.. 
Cmq stavo vedendo Italtherm, ma tu hai per caso questo controller? https://www.italtherm.it/hps/sistemi-di-controllo/

Sembra molto simile al mio, anche se io ho una Ferroli..

**Reply by lucky5822** [reply]
Ciao @Marcopal71  so che è passato un po' da questo messaggio ma sei mica riuscito a recuperare i registri da italtherm? Io vorrei provare a collegarla ho già l'adattatore modbus ma non trovo i registri di italtherm.

**Comment by Bergoglio** [comment]
Buongiorno,
provo ad aiutarti per i ventilconvettori (premesso ovviamente che non sono un luminare di Home Assistant, anzi)
Qui [https://www.home-assistant.io/integrations/modbus/#hvac_mode_register](url) viene specificato che puoi gestire diversi setpoint
In linea generale dovresti poter creare in configuration.yaml un listato dove radunare tutti i registri da leggere/scrivere per ogni singolo ventilconvettore.
Dopo, presumo, vorrai scrivere questi registri a partire da un termostato, corretto ?
Il termostato generico di Home Assistant permette di gestire o solo il caldo o solo il freddo. Esistono componenti aggiuntivi installabili via HACS che creano un termostato caldo + freddo, questo è il primo che mi viene a tiro [https://github.com/swingerman/ha-dual-smart-thermostat](url)
In questo modo gestisci tutto in uno e dimezzi il numero di termostati necessari oltre ad eliminare il rischio di far partire lo stesso ventilconvettore in riscaldamento e raffreddamento in contemporanea (problema molto probabile se hai un impianto a 4 tubi ovvero con un circuito per il caldo ed un altro circuito per il freddo, come me).
Sicuramente però dovrai creare un'automazione per impostare/scrivere i ventilconvettori perché i termostati che ho trovato lavorano tutti su uno switch. Io creerei uno switch virtuale (Impostazioni > Dispositivi e servizi > Aiutanti) alla cui attivazione una automazione accende / spegne il generatore termico + accende / spegne il ventilconvettore e gli scrive i registri.
Questi sono i primi concetti generali, tanto per scaldare i motori; ovviamente possiamo approfondire qui i vari aspetti.
Spero di aver centrato la questione.

P.S. : in quali linguaggi programmi ?

**Reply by crug80** [reply]
Grazie per l'interesse ed i consigli. Rispetto al post ho studiato un altro po'..  Il componente climate di modbus ha 2 limitazioni: non gestisce le ventole e non ha possibilità di gestire indirizzi diversi per i setpoint dedicati alle diverse modalità HVAC. Il primo problema sembra sia risolto, poiché dando un'occhiata al github nella prossima release dovrebbero esserci le ventole e le velocità. Rimane il problema dei registri che sarei quasi tentato di affrontare contribuendo al progetto. Devo studiare un altro po per vedere se è una cosa alla mia portata. 
Il concetto di termostato, attualmente non l'ho ancora affrontato e, sinceramente, non sono riuscito a capire se mi occorre. Nel senso che ho già creato delle entità climate che mi gestiscono i diversi temoconvettori ed in cui posso impostare la modalità di funzionamento (caldo/freddo/ventilazione) e temperatura (ora solo per il caldo, vedi problema sopra), quindi non ho ben afferrato in cosa potrebbe essermi utili un termostato. Forse per la PdC? Anche lì non saprei, perché il mio impianto è progettato per scaldare/freddare l'acqua in un puffer, quindi il generatore lavora a prescindere dalle temperature di casa che mi gestisco con i terminali. Se mi aiuti a capire se e perché è necessario un termostato, quando già i vari climate hanno tutto ciò che mi serve te ne sarei grato. 

Per la domanda nel p.s, sviluppo software dedicati a specifici settori in C++/C#

**Reply by Bergoglio** [reply]
Scusa, le entità climate che gestiscono i ventilconvettori non sono già dei generic thermostat ?
Puoi postare una schermata di queste entità che così capisco bene cosa sono ?

**Reply by crug80** [reply]
![immagine](https://github.com/indomus/forum/assets/41435902/0da16d02-5154-4822-b724-508a0d82b408)

Non far caso alla grafica, perché per ora mi sto concentrando sulle funzionalità e basta..

**Reply by Bergoglio** [reply]
Infatti come pensavo, hai impostato un generic thermostat per ogni ventilconvettore. Queste entità climate non sono specifiche di modbus, possono gestire una caldaia, una stufa elettrica, etc. etc. etc.
Il problema è che non comunicano (od almeno non mi pare) se stanno gestendo il riscaldamento oppure il raffrescamento e comunque devi dirgli tu manualmente se vuoi riscaldare o raffreddare.
Qui ci sono due soluzioni :
1) sostituire questi termostati con un dual thermostat da HACS come quello che ti ho postato e così sei sicuro di non accendere in contemporanea riscaldamento con raffreddamento, ma i componenti che avevo visto difettavano di altri elementi tipo il controllo finestre aperte che devi poi implementare con le automazioni (e non è il massimo)
2) raddoppiare i termostati creandone per ogni ventilconvettore uno di riscaldamento ed uno di raffreddamento, facendo sempre manualmente attenzione a non sovrapporre le temperature.
Io ho optato per la seconda strada, poi ho implementato un termostato da HACS chiamato Versatile Thermostat perché ha alcune funzioni accessorie utili. Questo componente è già valido ma volevo far mettere mano al codice da mia figlia che il Python lo mastica; il problema è che mia figlia non ha tutto questo tempo (va ancora al Liceo) e conosce il Python (con tanto di certificazione americana) ma per il data analysis e non per scrivere software applicativo (quindi deve sperimentare ed imparare).

**Reply by crug80** [reply]
Scusa, ma qualcosa ancora non mi è chiaro (è da poco che mi diletto con HA). 
Il componente è modbus, nel senso che l'implementazione modbus eredita ciò che un termostato fa ma lo applica al protocollo relativo e quindi automatizza le operazioni di scrittura/lettura registri. 
Cosa dovrebbe fare per gestire le 2 modalità? 
Nel senso, in base al tipo di accensione, il device (PdC o termoconvettore) si accende nella relativa modalità: raffrescamento/riscaldamento. E' vero, non posso memorizzare temperature predefinite per le modalità, ma posso impostarle in modo manuale/automatizzato e posso sapere dai suoi attributi cosa sta facendo: off, heat etc.. 
![immagine](https://github.com/indomus/forum/assets/41435902/9d8b295c-aac5-4097-a64b-8088e8f00f4c)

Perché dovrei avere 2 termostati, oppure 1 dual? 
Non riesco a capire che vantaggio avrei rispetto a questo o meglio, a parte il discorso dei registri, che limitazioni ha questo componente.

**Reply by Bergoglio** [reply]
Modbus è lo scrivano, prende i parametri dal termostato e li scrive sulla macchina. Il problema è che se il termostato non dice una temperatura target per il riscaldamento ed una temperatura target per il raffreddamento, come fa modbus a decidere in quale registro scrivere la temperatura ?
Potresti ovviare creando una automazione che dice :
trigger
ogni volta che temperature (target) cambia
azione
se hvac_modes è heat chiama servizio modbus scrivi su registro temperatura caldo valore temperature dal termostato altrimenti chiama servizio modbus scrivi su registro temperatura freddo valore temperature dal termostato
Questo in linea di principio ultra semplificato

**Reply by Bergoglio** [reply]
Ancora meglio nell'azione sarebbe inserire un choose e non if - then

**Reply by Bergoglio** [reply]
Comunque, per esperienza, ci sono meno intoppi se hai un termostato con due temperature target esposte (od in assenza un termostato solo per il caldo ed un secondo solo per il freddo) e due semplici automazioni, una per scrivere nel registro della temperatura riscaldamento ed una seconda per scrivere nel registro della temperatura raffrescamento.
Più le automazioni sono semplici e meno rischi di problemi hai.
Ancora meno se questi passaggi sono eseguiti da pezzi aggiuntivi di software (in Python) come quelli che trovi in HACS

**Reply by crug80** [reply]
Afferrato il concetto. Probabilmente stavamo dicendo la stessa cosa, ma mi esprimevo male io. 
Cmq, per la specifica esigenza, non mi interesserebbe avere 2 temperature per le modalità, ma 2 registri in cui scrivere la temperatura target a seconda della modalità. Attualmente è questo ciò che manca al componente.  
Per adesso il problema non si pone, poiché si accende sempre e solo in modalità caldo e quindi all'interno del file di configurazione ho scritto il registro relativo a questa modalità. Se fino all'estate non sono riuscito a cambiarne il comportamento, allora nella peggiore delle ipotesi cambio a mano il registro nel file di configurazione e per tutta l'estate starò a posto. 
Ovvio che cercherò di ovviare o provando a contribuire al progetto e quindi proporre una modifica al componente, oppure provando a customizzarlo per la mia esigenza. Devo studiarmi bene come funziona il processo di modifica/validazione del codice, poi mi regolerò.
Ma al momento la mia priorità è il funzionamento di casa! 

Durante le festività mi studio un po' la sintassi di Python e provo a metterci le mani. 

Intanto grazie per la pazienza e per il piacevole scambio di idee.

**Reply by Bergoglio** [reply]
Puoi automatizzare la scrittura sui due registri nel ventilconvettore con l'automazione che ti suggerivo sopra.
Avendo anch'io un sistema a ventilconvettori (leggermente diversi nel sistema di gestione) ho dovuto affrontare il medesimo problema.
Teniamoci in contatto perché qualche idea in merito l'ho elaborata, ma mi mancano alcuni pezzi che dovresti avere tu.

**Reply by crug80** [reply]
vorrei evitare automazioni ed usare il componente pulito di modbus perché sul forum di HA alcuni utenti hanno lamentato difficoltà nell'utilizzo di componenti diversi come ad esempio il [climate_templat](https://github.com/jcwillox/hass-template-climate)e che avrebbe fatto anche al caso mio.. 
E' probabile che essendo la parte modbus scritta dallo stesso autore della libreria pymodbus inclusa in HA, il suo componente climate sia più ottimizzato e quindi soffra di minor problemi rispetto a lavorare direttamente con i servizi di scrittura/lettura tramite automazioni. 

Per il resto, mi trovi qui sul forum o se c'è qualche modo di inviare messaggi, contattami pure senza problemi!

**Reply by Bergoglio** [reply]
Ok, ho trovato i tuoi recapiti sul tuo sito. I miei te li invio sull'e-mail, altrimenti potrebbero giustamente protestare qui sul forum.

Sono curioso di vedere come agisci sul modbus (magari con un collegamento remoto via Anydesk o Google Meet?).

**Comment by Marcopal71** [comment]
Ciao, purtroppo non mi hanno mai risposto. Se riesci tu fammi sapere.

Il giorno mer 27 nov 2024 alle 19:13 lucky5822 ***@***.***>
ha scritto:

**Reply by lucky5822** [reply]
Domani sento in italtherm e ti dico. Intanto ho guardato ed essendo la italtherm un clone midea credo che i registri siano gli stessi. Gli indirizzi midea li ho trovati.

**Comment by Marcopal71** [comment]
Se puoi girarmeli faccio qualche prova. Il mio inveter del fotovoltaico
l’ho già interfacciato in modbus e integrato in home assistant

Il giorno mer 27 nov 2024 alle 20:50 lucky5822 ***@***.***>
ha scritto:

**Reply by lucky5822** [reply]
Trovi i registri da pag 48 in avanti.

https://shop.systemair.com/upload/assets/OPERATION_MANUAL_EN_20210407_153552287.PDF?f39b69c9

Tu che modello hai? Io ho una hydrablock C. Se guardi al link che ti ho inviato, il manuale e identico a quello italtherm e il controllo (pannellino) è lo stesso. Siccome il controllo mostra i parametri che recupera da modbus credo proprio che i registri siano quelli. Fammi sapere che ne pensi

**Comment by Marcopal71** [comment]
Io ho una
HYDRABLOCK Hybrid TOP 35K/12M

Il giorno mer 27 nov 2024 alle 20:58 lucky5822 ***@***.***>
ha scritto:

**Reply by lucky5822** [reply]
Mi mandi per favore un esempio di come configurare i sensori in home assistant con modbus? Si impostano in configuration.yaml giusto?

**Reply by lucky5822** [reply]
Ieri ho fatto una prova con quei registri ma a me da sempre unavailable il sensore. Te sei riuscito a fare qualcosa?

**Comment by Marcopal71** [comment]
Ciao, purtroppo in questi giorni non riesco a fare prove, ma ti scrivo
quello che ho fatto con l'inverter.

Prima va definito il protocollo di comunicazione, poi i vari sensori, letti
singolarmente o in gruppi che poi vanno spacchettati

modbus:
  - name: "solis"
    type: serial
    method: rtu
    port: /dev/ttyUSB0
    baudrate: 9600
    stopbits: 1
    bytesize: 8
    parity: N


    sensors:
      - name: solismodbus_33073_33084
        slave: 1
        address: 33073
        count: 12
        input_type: input
        data_type: custom
       #   lazy_error_count: 5
        scan_interval: 2
        structure: ">HHHHHHlll"

      - name: solismodbus_33126_33152
        slave: 1
        address: 33126
        count: 27
        input_type: input
        data_type: custom
      # lazy_error_count: 5
        scan_interval: 10
        structure: ">LHHlHHhHHHHHHHhHHHHHHll"

      - name: solismodbus_33161_33180
        slave: 1
        address: 33161
        count: 20
        input_type: input
        data_type: custom
       # lazy_error_count: 5
        scan_interval: 60
        structure: ">LHHLHHLHHLHHLHH"

      - name: solismodbus_33251_33286
        slave: 1
        address: 33251
        count: 36
        input_type: input
        data_type: custom
       # lazy_error_count: 5
        scan_interval: 10
        structure: ">HHHHHHllllllllllllhHLL"

      - name: solismodbus_33029_33058
        slave: 1
        address: 33029
        count: 30
        input_type: input
        data_type: custom
    #    lazy_error_count: 5
        scan_interval: 10
        structure: ">LLLHHLLHHHHHHHHHHHHHHHHL"

      - name: solismodbus_33093_temperature
        slave: 1
        address: 33093
       # count: 1
        input_type: input
        data_type: uint16
       # lazy_error_count: 5
        scan_interval: 60

      - name: 33139_SOC_Batt
        slave: 1
        address: 33139
      #  count: 1
        input_type: input
        data_type: uint16
        state_class: measurement
        device_class: battery
        unit_of_measurement: "%"
       # lazy_error_count: 5
        scan_interval: 10

      - name: 33140_SOH_Batt
        slave: 1
        address: 33140
      #  count: 1
        input_type: input
        data_type: uint16
     #   lazy_error_count: 5
        scan_interval: 10

Il giorno gio 28 nov 2024 alle ore 10:44 lucky5822 ***@***.***>
ha scritto:

**Reply by lucky5822** [reply]
Ciao sì nel frattempo ho fatto un po' di prove anche io e ti confermo che la lista indirizzi modbus che ti ho linkato è corretta. Sono riuscito ad inserire e leggere tutti i sensori che mi interessavano. Per il momento funziona tutto.
